FIAT/christiansen_hu.py

```diff
@@ -0,0 +1,81 @@
+# This file is part of FIAT (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+#
+# Written by Pablo D. Brubeck (brubeck@protonmail.com), 2024
+
+# This is not quite Christiansen-Hu, but it has 2*dim*(dim+1) dofs and includes
+# dim**2-1 extra constraint functionals.  The first (dim+1)**2 basis functions
+# are the reference element bfs, but the extra dim**2-1 are used in the
+# transformation theory.
+
+from FIAT import finite_element, polynomial_set
+from FIAT.bernardi_raugel import BernardiRaugelDualSet
+from FIAT.quadrature_schemes import create_quadrature
+from FIAT.reference_element import TETRAHEDRON
+from FIAT.macro import CkPolynomialSet, WorseyFarinSplit
+
+import numpy
+
+
+def ChristiansenHuSpace(ref_el, degree, reduced=False):
+    """Return a basis for the Christianse-Hu space
+    set(v in C0 P1(WF)^d : div(v) = 0) + P_0 x if reduced = True, and
+    this space is agumented with rotated facet bubbles if reduced = False."""
+    sd = ref_el.get_spatial_dimension()
+    ref_complex = WorseyFarinSplit(ref_el)
```

Comment:
I think of Worsey-Farin as a 3d thing.  Is it defined on triangles (or does it reduce to something else)?

Reply:
You are right, it is a 3D thing only, but we could extend it to any dimension by saying that we recursively applying barycentric refinement on entities of codim <= 1. This Reduces to PS6 on the triangle, and Alfeld on the interval.